libtransmission/tr-udp.cc

```diff
@@ -331,70 +327,126 @@ void tr_udpInit(tr_session* ss)
 
     if (tr_globalIPv6(nullptr) != nullptr)
     {
-        rebind_ipv6(ss, true);
+        rebind_ipv6(true);
     }
 
-    if (ss->udp6_socket != TR_BAD_SOCKET)
+    if (udp6_socket_ != TR_BAD_SOCKET)
     {
-        ss->udp6_event = event_new(ss->eventBase(), ss->udp6_socket, EV_READ | EV_PERSIST, event_callback, ss);
+        udp6_event_ = event_new(session_->eventBase(), udp6_socket_, EV_READ | EV_PERSIST, event_callback, session_);
 
-        if (ss->udp6_event == nullptr)
+        if (udp6_event_ == nullptr)
         {
             tr_logAddWarn(_("Couldn't allocate IPv6 event"));
         }
     }
 
-    tr_udpSetSocketBuffers(ss);
-
-    tr_udpSetSocketTOS(ss);
+    set_socket_buffers();
+    set_socket_tos();
 
-    if (ss->allowsDHT())
+    if (session_->allowsDHT())
     {
-        tr_dhtInit(ss);
+        tr_dhtInit(session_);
     }
 
-    if (ss->udp_event != nullptr)
+    if (udp_event_ != nullptr)
     {
-        event_add(ss->udp_event, nullptr);
+        event_add(udp_event_, nullptr);
     }
-
-    if (ss->udp6_event != nullptr)
+    if (udp6_event_ != nullptr)
     {
-        event_add(ss->udp6_event, nullptr);
+        event_add(udp6_event_, nullptr);
     }
 }
 
-void tr_udpUninit(tr_session* ss)
+void tr_session::tr_udp_core::uninit()
 {
-    tr_dhtUninit(ss);
+    tr_dhtUninit(session_);
+
+    if (udp_socket_ != TR_BAD_SOCKET)
+    {
+        tr_netCloseSocket(udp_socket_);
+        udp_socket_ = TR_BAD_SOCKET;
+    }
 
-    if (ss->udp_socket != TR_BAD_SOCKET)
+    if (udp_event_ != nullptr)
     {
-        tr_netCloseSocket(ss->udp_socket);
-        ss->udp_socket = TR_BAD_SOCKET;
+        event_free(udp_event_);
+        udp_event_ = nullptr;
     }
 
-    if (ss->udp_event != nullptr)
+    if (udp6_socket_ != TR_BAD_SOCKET)
     {
-        event_free(ss->udp_event);
-        ss->udp_event = nullptr;
+        tr_netCloseSocket(udp6_socket_);
+        udp6_socket_ = TR_BAD_SOCKET;
     }
 
-    if (ss->udp6_socket != TR_BAD_SOCKET)
+    if (udp6_event_ != nullptr)
     {
-        tr_netCloseSocket(ss->udp6_socket);
-        ss->udp6_socket = TR_BAD_SOCKET;
+        event_free(udp6_event_);
+        udp6_event_ = nullptr;
     }
 
-    if (ss->udp6_event != nullptr)
+    if (udp6_bound_ != nullptr)
+    {
+        free(udp6_bound_);
+        udp6_bound_ = nullptr;
+    }
+}
+
+void tr_session::tr_udp_core::sendto(char const* buf, size_t buflen, struct sockaddr const* to, socklen_t const tolen) const
+{
+    ssize_t ret = 0;
+    std::array<char, std::max(INET_ADDRSTRLEN, INET6_ADDRSTRLEN) + 1> peer = {};
+
+    if (to->sa_family == AF_INET)
+    {
+        if (udp_socket_ != TR_BAD_SOCKET)
+        {
+            ret = ::sendto(udp_socket_, buf, buflen, 0, to, tolen);
+        }
+        else
+        {
+            ret = -1, errno = EBADF;
```

Comment:
Here and below on 428 and 441

```suggestion
            ret = -1;
            errno = EBADF;
```